mlflow/transformers/__init__.py

```diff
@@ -2208,8 +2243,14 @@ def trim_input(data_in, data_out):
                 data_out = data_out[len(data_in) :].lstrip()
                 if data_out.startswith("A:"):
                     data_out = data_out[2:].lstrip()
-                # If the user has indicated to remove newlines and extra spaces from the generated
-                # text, replace them with a single space.
+
+            # only return generated text if a prompt template was set.
+            # this is to avoid leaking system prompts.
+            if self.prompt_template and data_out.startswith(data_in):
+                data_out = data_out[len(data_in) :].lstrip()
+
+            # If the user has indicated to remove newlines and extra spaces from the generated
+            # text, replace them with a single space.
```

Comment:
## Behavior Questions
this part trims out the prompt if the prompt template was set. this is in order to avoid leaking system prompts. i think this is the right thing to do, but my concern is that it will be confusing for model creators who maybe expect the full text to be returned (it's the [default behavior](https://huggingface.co/docs/transformers/main_classes/pipelines#transformers.TextGenerationPipeline.__call__.return_full_text) on text generation pipelines). a couple questions:

1. we should definitely document this, but should we also print a warning during save time to notify users about this behavior?
2. is there any case where we don't want to do this? should we make it configurable?

## Technical Questions
i wasn't sure if this is the right place to put this—however it looks like a post-processing function for `TextGenerationPipeline`s, so it felt right. to see where this gets called, see line 1907 in this file (expand the above block downward).

i also considered forcing `self.model_config` to have `return_full_text=False` ([link to transformers code](https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/text_generation.py#L179)) if the prompt template is set and the pipeline is `text-generation`. i think this is another valid way to approach the prompt trimming problem, but seems kind of [equivalent](https://github.com/huggingface/transformers/blob/main/src/transformers/pipelines/text_generation.py#L308-L310).

Reply:
curious why we should avoid leaning system prompts

Reply:
I'm in favor of using the feature that exists within transformers for the sole purpose of avoiding package version conditional logic. 
However, that being said, I do think we need a try/catch around this to fail permissively if the trimming function within transformers doesn't work. We should return the full payload if we can't trim it rather than throwing an unfixable exception for the users. 

As far as configuration goes, if we specify `return_full_text` on `mlflow.transformers.load_model()` (as a kwarg, since we do pass-through of load kwargs), is that behavior configurable? 
If so, I'd say that's our solution: document that this is something that can be done and leave it to the transformers library to handle how this is done at load time. 

Can we verify this behavior so that we can avoid creating any of this parsing logic?

Reply:
> curious why we should avoid leaking system prompts.

i think that in some cases it isn't a big deal (e.g. if the system prompt is just some generic guidance for the model). however in other situations, model creators may have spent a long time fine-tuning a proprietary prompt that they want to keep secret, or the prompt may have some internal company info that they might not want to leak.

we can probably make a case that we don't actually need to do this, and if people care about leaking the original prompt then they can write a custom pyfunc to strip out the prompt themselves—it's hard to say what the right experience should be, but intuitively it feels to me like the system prompt should be stripped out. cc @dbczumar do you have any input here?

> Can we verify this behavior so that we can avoid creating any of this parsing logic?

Will follow up on this! I think it makes sense, and might be a better user experience. We can default the pipeline config to be `return_full_text=False`, but let users override it through the model_config if they want.

Reply:
I think it's fair to trim out the system prompt - it likely doesn't have any meaning to the end user who's calling the model anyway. Shouldn't transformers provide a flag / option for this, as @BenWilson2 suggests?

Reply:
Yup they do! Investigating using that instead